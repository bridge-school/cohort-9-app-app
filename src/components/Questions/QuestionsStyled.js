import styled from "styled-components";

export const Title = styled.h1`
  width: 100%;
  text-align: left;
  font-size: 35px;
  font-weight: 800;
  color: #333;
  margin-bottom: 40px;
`