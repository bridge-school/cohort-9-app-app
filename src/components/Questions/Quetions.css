.Question-deleteBtn {
  /* display: flex; */
  align-self: center;
}

.Question-checkbox {
  align-self: center;
}
