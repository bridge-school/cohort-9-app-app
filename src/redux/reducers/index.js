import { combineReducers } from "redux";
import datesReducer from "./datesReducer";
import adminFormReducer from "./adminFormReducer";
// import studentFormReducer from "./studentFormReducer";
import allCohortAppsReducer from "./allCohortAppsReducer";
// import cohortAppReducer from "./cohortAppReducer";
import getCohortApplicationByIdReducer from "./getCohortApplicationByIdReducer"
import studentSubmissionReducer from "./studentSubmissionReducer"

const rootReducer = combineReducers({
  dates: datesReducer,
  cohortInfo: adminFormReducer,
  cohortApplication: getCohortApplicationByIdReducer,
  studentSubmission: studentSubmissionReducer,
  apps: allCohortAppsReducer,

});

export default rootReducer;
