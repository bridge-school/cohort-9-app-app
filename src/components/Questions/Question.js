import React from "react";
import { Form, Checkbox } from 'semantic-ui-react'
import "./Quetions.css";

const Question = ({
  question,
  index,
  onPromptChange,
  onOptionsChange,
  onTypeChange,
  onIsRequiredChange,
  onDelete
}) => {
  const qNumber = index + 1;
  const optionsString = question.options ? question.options.join(', ') : ''
  const QUESTION_TYPE_OPTIONS = [
    { key: 'short-answer', text: 'Short Answer', value: 'short-answer'},
    { key: 'drop-down', text: 'Dropdown', value: 'drop-down'},
    { key: 'check-box', text: 'Checkboxes', value: 'check-box'},
    { key: 'paragraph', text: 'Paragraph', value: 'paragraph'}
  ]

  return (
    <>
      <Form.Group>
        <Form.Input
          id={`q${index}__name`}
          label={{ children: `Question #${qNumber}`, htmlFor: `q${index}__name` }}
          onChange={(e, {value}) => {
            onPromptChange(index, value);
          }}
          value={question.prompt}
          width={8}
        />
        <Form.Select
          fluid
          id={`q${index}__type`}
          label={{ children: `Question #${qNumber} Type`, htmlFor: `q${index}__type` }}
          options={QUESTION_TYPE_OPTIONS}
          onChange={(e, {value}) => {
            onTypeChange(index, value);
          }}
          value={question.type}
          width={4}
        />
        <Form.Field
          id={`q${index}__req`}
          className="Question-checkbox"
          control={Checkbox}
          label={{ children: `Is Required`, htmlFor: `q${index}__req` }}
          onChange={(e, {checked: value}) => {
            onIsRequiredChange(index, value);
          }}
          checked={question.required}
          width={2}
        />
        <Form.Button
          className="Question-deleteBtn"
          content="X"
          type="button"
          onClick={() => {
            onDelete(index);
          }}
          width={2}
        />
      </Form.Group>
      {(question.type === "check-box" || question.type === "drop-down") &&
        <Form.Group>
          <Form.Input
            id={`q${index}__options`}
            label={{ children: `Question #${qNumber} type options`, htmlFor: `q${index}__options` }}
            onChange={(e, {value}) => {
              onOptionsChange(index, value);
            }}
            value={optionsString}
            placeholder="a, b, c, d"
            width={16}
          />
        </Form.Group>
      }
    </>
  );
};

export default Question;
