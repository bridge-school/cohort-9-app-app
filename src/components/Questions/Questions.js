import React from "react";
import { connect } from "react-redux";
import { Button, Container, Divider } from 'semantic-ui-react'
import Question from "./Question";
import { 
  addNewQuestion, 
  deleteQuestion,
  setQuestionPrompt, 
  setQuestionType, 
  toggleQuestionRequired, 
  setQuestionOptions, 
} from "../../redux/actions/adminFormActions";
import {Title} from "./QuestionsStyled";

const Questions = (props) => {
  // destructure props create a constant with questionsData
  const { questionsData } = props;

  const addNewQuestion = e => {
    e.preventDefault();
    props.addNewQuestion();
  };

  const handleQuestionPromptChange = (index, name) => {
    props.setQuestionPrompt(index, name);
  };

  const handleQuestionTypeChange = (index, type) => {
    props.setQuestionType(index, type);
  };

  const handleQuestionOptionsChange = (index, optionsString) => {
    let optionsArray = [];
    // input comes in as a string, we convert it 
    // into an array for storing in redux for easier future use
    if (optionsString) {
      optionsArray = optionsString
        .split(",")
        // remove whitespace around the option (if any)
        .map(option => option.trim());
    }
    props.setQuestionOptions(index, optionsArray);
  };

  const handleIsRequiredChange = index => {
    props.toggleQuestionRequired(index);
  };

  const handleDelete = index => {
    props.deleteQuestion(index);
  };
  
  return(
    <Container>
      <Divider hidden />
      <Title>Application Questions</Title>
      {questionsData.map((question, i) => {
        return (
          <Question
            key={`question_${question.timestampForKey}`}
            index={i}
            question={question}
            onPromptChange={handleQuestionPromptChange}
            onTypeChange={handleQuestionTypeChange}
            onOptionsChange={handleQuestionOptionsChange}
            onIsRequiredChange={handleIsRequiredChange}
            onDelete={handleDelete}
          />
        );
      })}
      <Divider hidden />
      <Container textAlign='center'>
        <Button basic color='green' size='big' onClick={addNewQuestion}>Add new question</Button>
      </Container>
    </Container>
  );
};

const mapStateToProps = state => {
  return {
    questionsData: state.cohortInfo.questionsData
  };
};

const mapDispatchToProps = dispatch => ({
  addNewQuestion: () => dispatch(addNewQuestion()),
  setQuestionPrompt: (questionIndex, promptText) => {
    dispatch(setQuestionPrompt(questionIndex, promptText));
  },
  setQuestionType: (questionIndex, options) => {
    dispatch(setQuestionType(questionIndex, options));
  },
  toggleQuestionRequired: (questionIndex) => {
    dispatch(toggleQuestionRequired(questionIndex));
  },
  setQuestionOptions: (questionIndex, optionsString) => {
    dispatch(setQuestionOptions(questionIndex, optionsString));
  },
  deleteQuestion: questionIndex => dispatch(deleteQuestion(questionIndex))
});

export default connect(mapStateToProps, mapDispatchToProps)(Questions);
