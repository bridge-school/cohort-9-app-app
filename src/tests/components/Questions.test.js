import React from "react";
import renderer from "react-test-renderer";
import Questions from "../../components/Questions/Questions";
import { Provider } from "react-redux";
import { store } from "../../redux/store";

describe("Questions", () => {
  test("Should render without questions properly", () => {
    const questions = [];

    const tree = renderer
      .create(
        <Provider store={store}>
          <Questions questionsData={questions} />
        </Provider>
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot();
  })

  test("Should render questions properly", () => {
    const questions = [
      {
        timestampForKey: '1575845632',
        prompt: 'question 1',
        options: ['a', 'b'],
        type: 'check-box',
        required: true
      },
      {
        timestampForKey: '1575845632',
        prompt: 'question 2',
        options: ['a', 'b', 'c', 'd'],
        type: 'drop-down',
        required: false,
      },
      {
        timestampForKey: '1575845632',
        prompt: 'question 3',
        type: 'paragraph',
        required: true,
      },
      {
        timestampForKey: '1575845632',
        prompt: 'question 4',
        type: 'short-answer',
        required: true,
      }
    ];
    
    const tree = renderer
      .create(
        <Provider store={store}>
          <Questions questionsData={questions} />
        </Provider>
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot();
  })
})