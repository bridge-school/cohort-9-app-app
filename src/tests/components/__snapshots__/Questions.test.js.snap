// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions Should render questions properly 1`] = `
<div>
  <h1>
    Application Questions
  </h1>
  <button
    onClick={[Function]}
  >
    Add new question
  </button>
</div>
`;

exports[`Questions Should render without questions properly 1`] = `
<div>
  <h1>
    Application Questions
  </h1>
  <button
    onClick={[Function]}
  >
    Add new question
  </button>
</div>
`;
